[SofaKernel][SofaCore][SofaLoader][SofaGeneralLoader] SOFA with callbacks

marques-bruno opened the pull request:
This PR is based on #1406 and #1407 and uses the callback mechanism on some SOFA components (mainly Mesh loaders)
It also takes advantage of the newly added "isDirectory" method in DataFileName in AddResourceRepository

It also:
- Removes the isPersistent flag from datafields in MeshLoaders (because this flag should only be used to set application-specific behaviors: in runSofa2, this flag is used to determine whether the datafield should be recorded when saving the scene for instance)
- takes advantage of the getAccessors introduced by @damienmarchal  in #1254 
- Clears log messages from components withing update callbacks if the componentState changes to "Valid"
- Sets the dynamic datafields created on-the-fly in the MeshVTKLoader depending on the mesh file content as outputs of the component (to guarantee they're update when the component's inputs change
- Checks for the validity of binary STL files using `isBinarySTLValid` for better error handling (see comment in the code l.103 in MeshSTLLoader.cpp)
- Removes the autoLink flag from fileMesh in VisualModelImpl

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

marques-bruno committed:
Using UpdateCallbacks to update internal component data & states

marques-bruno committed:
Merge branch 'add_directoryDataFileNames' into sofaWithCallbacks

marques-bruno committed:
Removing AutoLink flag from fileMesh field

damienmarchal commented on SofaKernel/modules/SofaCore/src/sofa/core/DataTracker.cpp:
```diff
@@ -63,15 +64,15 @@ void DataTracker::clean()
+void DataTrackerDDGNode::addOutputs(std::initializer_list<sofa::core::objectmodel::DDGNode*> datas)
 {
-    for(sofa::core::objectmodel::BaseData* d : datas)
+    for(sofa::core::objectmodel::DDGNode* d : datas)
```
I'm not sure of this part....A data tracker is supposed to track data...otherwise name that a DDGNodeTracker :)

marques-bruno replied:
Well it's actually a DataTrackerDDGNode... :D but yes I understand your point.
Having DataTrackerDDGNode register DDGNodes instead of BaseData will let us track nodephysics::Links (which are built upon the DDGNode API). But since the DDG Links are not part of this pull request I will remove this change from this PR

marques-bruno replied:
Taken into consideration in last commit (also applied on #1406 which introduces the code in question)

damienmarchal commented on SofaKernel/modules/SofaCore/src/sofa/core/loader/MeshLoader.cpp:
```diff
@@ -83,18 +83,6 @@ MeshLoader::MeshLoader() : BaseLoader()
     d_transformation.setAutoLink(false);
     d_transformation.setDirtyValue();
 
-    d_positions.setPersistent(false);
```
This is breaking, but this is really the way to go..anyway un the current state of the sofa code, saving is just un-unsable (and unused).

marques-bruno replied:
Good point. I set the breaking label on the PR

marques-bruno committed:
ADD: Check when setting value to guarantee consistency between value and isDirectory + apply same system to DataFileNameVector

marques-bruno committed:
Merge branch 'add_updateCallback' into sofaWithCallbacks

marques-bruno committed:
Merge branch 'add_updateCallback' into sofaWithCallbacks

marques-bruno commented on SofaKernel/modules/SofaBaseUtils/AddResourceRepository.cpp:
```diff
@@ -68,21 +73,39 @@ void BaseAddResourceRepository::parse(sofa::core::objectmodel::BaseObjectDescrip
+    Inherit1::parse(arg);
+
+    if (parse())
+        m_componentstate.setValue(sofa::core::objectmodel::ComponentState::Valid);
```
SetValue on componentState should not be required: `bool parse()` calls getValue on the m_repositoryPath, which in turn will call the updateCallback, which himself sets the component State. Thus setting it again would be useless

marques-bruno commented on SofaKernel/modules/SofaBaseVisual/VisualModelImpl.cpp:
```diff
@@ -221,6 +221,7 @@ VisualModelImpl::VisualModelImpl() //const std::string &name, std::string filena
 
     //material.setDisplayed(false);
     addAlias(&fileMesh, "fileMesh");
+    fileMesh.setAutoLink(false);
```
to remove

epernod reviewed (dismissed):
move generic addUpdateCallback inside MeshLoader parent class.
- position
- normals
- edges
- triangles
- quads
etc.. see Data declared in MeshLoader

marques-bruno commented on SofaKernel/modules/SofaLoader/MeshVTKLoader.cpp:
```diff
@@ -552,6 +554,8 @@ bool MeshVTKLoader::setInputsData()
 
         BaseData* basedata = reader->inputPointDataVector[i]->createSofaData();
         this->addData(basedata, dataname);
+        m_internalEngine["filename"].addOutput(basedata);
```
Add m_internalEngine as private & add a method `addOutputToCallback("callbackName", data)`

marques-bruno committed:
Remove the use of callback engines accepting multiple callback functions

marques-bruno committed:
refactoring of AddResourceRepository::parse()

marques-bruno committed:
Merge branch 'add_updateCallback' into sofaWithCallbacks

marques-bruno commented:
> move generic addUpdateCallback inside MeshLoader parent class.
> 
>     * position
> 
>     * normals
> 
>     * edges
> 
>     * triangles
> 
>     * quads
>       etc.. see Data declared in MeshLoader

Hi @epernod ! After reviewing my code a bit more, I now remember why MeshLoader doesn't contain a "generic" update callback: there's no "load" method in MeshLoader, only a set of datafield definitions, then used by other components subclassing it. thus a callback function would not be able to perform any loading operation on input changes. I agree that it is sub-optimal to have to add a specific callback for each meshLoaders (although they have custom behaviors anyway) but as-is I don't see how to better the current system. Ideally, we could imagine having a pure virtual "load" method overridden in the subclasses, which would then be callable from a callbackEngine, but that would probably go out of the scope of this PR.

marques-bruno commented:
poking @hugtalbot as requested, now that changes have been taken into consideration. We can consider closing #1406 when we merge this one. Still waiting for @epernod's response to my last comment before taking further action

marques-bruno committed:
ADD: usage of addOutputToCallbackEngine in MeshVTKLoader

marques-bruno committed:
Merge remote-tracking branch 'SofaDefrost/fixCxxVisibility' into sofaWithCallbacks

marques-bruno commented:
> Hi @marques-bruno wasn't it merging the #1407 into this one (#1408) ?
> 
> My notes on this PR were: Rebase #1407 in #1408.
> SetValue on componentState should not be required: bool parse() calls getValue on the m_repositoryPath, which in turn will call the updateCallback, which itself sets the component State. Thus setting it again would be useless.
> fileMesh to remove

That's what's been done here no? Both 1406 & 1407 are merged into this PR (see branch diffs), the parsing code has been updated in AddResourceRepository & fileMesh has been removed.
To me this should be flagged ready, as soon as @epernod validates

marques-bruno committed:
Merge branch 'sofaWithCallbacks' of github.com:SofaDefrost/sofa into sofaWithCallbacks

marques-bruno committed:
Update DataTracker.cpp

marques-bruno committed:
Update DataTracker.cpp

marques-bruno commented:
[ci-build][with-all-tests]

marques-bruno committed:
Merge branch 'add_updateCallback' into sofaWithCallbacks

marques-bruno committed:
Merge branch 'add_directoryDataFileNames' into sofaWithCallbacks

marques-bruno committed:
Update callbacks & datafilenames on loaders & resourceRepository

marques-bruno committed:
Merge branch 'add_updateCallback' into sofaWithCallbacks

marques-bruno committed:
Adding delegate doClearBuffers & final method clearBuffers in MeshLoader + move updateCallback to MeshLoader

marques-bruno committed:
FIX small indent pblem caused by macro instruction

marques-bruno commented on SofaKernel/modules/SofaLoader/MeshVTKLoader.cpp:
```diff
@@ -979,7 +985,7 @@ bool XMLVTKReader::readFile(const char* filename)
     checkErrorMsg(datasetFormatStrTemp, "Dataset format not defined")
-    string datasetFormatStr = string(datasetFormatStrTemp);
+            string datasetFormatStr = string(datasetFormatStrTemp);
     VTKDatasetFormat datasetFormat;
```
Sharp eye, Mr. @hugtalbot !

marques-bruno committed:
Merge branch 'master' into sofaWithCallbacks

marques-bruno committed:
move clearBuffers from BaseLoader to MeshLoader

marques-bruno committed:
ADD delegate doLoad() to properly clear buffers

marques-bruno committed:
Merge branch 'master' into sofaWithCallbacks

marques-bruno committed:
Merge branch 'master' into sofaWithCallbacks

marques-bruno committed:
Merge branch 'sofaWithCallbacks' of github.com:SofaDefrost/sofa into sofaWithCallbacks

marques-bruno commented:
[ci-build][with-all-tests]

guparan commented:
@marques-bruno No need to call a build when you push, it's already triggered :wink:

marques-bruno committed:
ADD: addTriangle function taking writeOnlyAccessor

marques-bruno commented:
> @marques-bruno No need to call a build when you push, it's already triggered wink

sorry for the spam...: I often push twice, then cancel the 1st build and then the 2nd needs a manual trigger...

marques-bruno committed:
UpdateMeshSTEPLoader

marques-bruno committed:
FIX compilation MeshSTEPLoader & fix GMshLoader

marques-bruno committed:
FIX DDG redundancies in MeshGmshLoader

marques-bruno committed:
Fix AssimpLoader

marques-bruno committed:
Fix AssimpLoader

marques-bruno commented:
@sofa-framework/reviewers one more review?

marques-bruno committed:
FIX: redundant callback loops in DataCallbacks

marques-bruno committed:
FIX: TrackedData_testInsideComponent
Test was failing because callback inputs are clean but their dirtyOutputs flag was still dirty

marques-bruno committed:
Remove comment

marques-bruno committed:
Merge branch 'master' into sofaWithCallbacks

marques-bruno committed:
fix typo from master merge conflict

guparan commented:
[ci-build][with-all-tests][force-full-build]

marques-bruno committed:
:bug: use getAccessor in GridMeshCreator

Should fix scene errors in anisotropic2DFEM.scn
